tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp

```diff
@@ -243,76 +248,279 @@ struct FileContent {
+  ASTUnitRef buildASTUnit(std::string &Error);
+
+  /// Retrieve the contents of all files needed for the compiler invocation to
+  /// build this AST. For files contained in \c Snapshots use the snapshot's
+  /// content. For all other files, consult the file system.
+  void findSnapshotAndOpenFiles(SmallVectorImpl<FileContent> &Contents,
+                                std::string &Error) const;
+
+  /// Transition the build operation to \p NewState, asserting that the current
+  /// state is \p ExpectedOldState.
+  void transitionToState(State NewState, State ExpectedOldState) {
+    assert(OperationState == ExpectedOldState);
+    OperationState = NewState;
+  }
+
+  /// Create a vector of text snapshots containing all files explicitly
+  /// referenced by the compiler invocation and a vector of buffer stamps of
+  /// those files. For all files that have a snapshot specified in \p
+  /// FixedSnapshots, that snapshot is used. For all other files \p FileSystem
+  /// will be consulted for the current contents of the file.
+  std::pair<std::vector<ImmutableTextSnapshotRef>, std::vector<BufferStamp>>
+  snapshotAndStampsForFilesInCompilerInvocation(
+      ArrayRef<ImmutableTextSnapshotRef> FixedSnapshots);
 
 public:
-  explicit ASTProducer(SwiftInvocationRef InvokRef)
-    : InvokRef(std::move(InvokRef)) {}
+  ASTBuildOperation(IntrusiveRefCntPtr<llvm::vfs::FileSystem> FileSystem,
+                    ArrayRef<ImmutableTextSnapshotRef> Snapshots,
+                    SwiftInvocationRef InvokRef, SwiftASTManagerRef ASTManager,
+                    std::function<void(void)> DidFinishCallback)
+      : InvokRef(InvokRef), FileSystem(FileSystem),
+        ASTManager(ASTManager), DidFinishCallback(DidFinishCallback) {
+    auto SnapshotsAndStamps = snapshotAndStampsForFilesInCompilerInvocation(Snapshots);
+    // const_cast is fine here. We just want to guard against modifying these
+    // fields later on. It's fine to set them in the constructor.
+    const_cast<std::vector<ImmutableTextSnapshotRef> &>(this->Snapshots) = SnapshotsAndStamps.first;
+    const_cast<std::vector<BufferStamp> &>(this->Stamps) = SnapshotsAndStamps.second;
+  }
+
+  ~ASTBuildOperation() {
+    assert(OperationState == State::Finished &&
+           "ASTBuildOperations should only be destructed once they have "
+           "produced an AST or are finished. Otherwise, some consumers might "
+           "not receive their callback.");
+  }
+
+  ArrayRef<ImmutableTextSnapshotRef> getSnapshots() const { return Snapshots; }
+
+  /// Returns true if the build operation has finished.
+  bool isFinished() {
+    llvm::sys::ScopedLock L(ConsumersAndResultMtx);
+    return Result.HasValue;
+  }
 
-  ASTUnitRef getExistingAST() {
-    // FIXME: ThreadSafeRefCntPtr is racy.
-    llvm::sys::ScopedLock L(Mtx);
-    return AST;
+  bool isCancelled() {
+    llvm::sys::ScopedLock L(ConsumersAndResultMtx);
+    return (Result.HasValue && Result.Cancelled) || *CancellationFlag;
```

Comment:
Is this called often? We could optimize a bit by checking CancellationFlag before taking the lock.

Reply:
Only when adding new consumers, so not particularly often, I’d say. I’d prefer to keep it this way for readability.